src/tensors/cpu/ruy_adapter.h

```diff
@@ -0,0 +1,534 @@
+/*
+ * This file follows intgemm and is a means of retrofitting ruy into the intgemm based wiring in
+ * `intgemm_interface.h`. ruy is an inference backend used in tensorflow and android deployment and
+ * has an optimized ARM backend for the multiply operations required. Optimized code for quantize,
+ * unquantize, transpose are added separately to connect the multiply library to marian.
+ */
+
+#pragma once
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include "ruy/platform.h"
+#include "ruy/system_aligned_alloc.h"
+
+#if RUY_PLATFORM_NEON
+#include <arm_neon.h>
+#endif
+
+namespace marian {
+namespace cpu {
+namespace integer {
+
+using Index = unsigned int;
+
+/*
+ * An AlignedVector is similar to intgemm's aligned allocations. Defined here
+ * independently because we are ignoring intgemm path entirely on ARM.
+ */
+template <class T>
```

Comment:
My question was more so, why do you need to allocate memory at all and when doing that, why do it outside of marian's framework?

The reason i'm asking this is because marian uses memory pool for managing memory allocations, which means that you only get a handful of calls to `malloc` (only when expanding the workspace/running out of memory of the current pool). This is by design because `malloc` is not only a very expensive call, as it goes to kernel space, but also implies a lock (`tcmalloc` makes the locking not as expensive, but still...). Allocating memory outside of the marian's memory pool and doing so repeatedly will carry significant performance penalty.

Reply:
I believe you mean `TensorAllocator` -> `Allocator`. The static large workspace allocation then lockless provides memory pieces to calling thread. The intgemm/ruy subsystem is (and should be) unaware of this. 

The allocation in question happens for `int32_t` accumulator, which is then converted to `float32_t`. (https://github.com/google/gemmlowp/issues/203). I think there might be a way to reuse the `float_32t` memory to write-out by in-place changes after `int32_t`? Maybe we can take that route.

Tangent: Intgemm integration appears to be doing allocations [outside the workspace](https://github.com/browsermt/marian-dev/blob/844800efccba6e670250caac1735ca2c8c8e508e/src/tensors/cpu/integer_common.h#L65-L67).

Reply:
The intgemm code you are referring to is called only once during model loading, it is not used during decoding,  and is used because at that point in time there is no graph instantiated that can be used.

During decoding, intgemm doesn't make use of any temporary memory allocations, and if you really have to, you have to make sure they don't happen at every GEMM operation.

Reply:
[`b271b70` (#79)](https://github.com/browsermt/marian-dev/pull/79/commits/b271b70bfe7337a4cacca789cc657dc020968cc2)

Appears to be passing tests at https://github.com/jerinphilip/MozIntGemm/pull/31. Using the externally supplied float* from marian's allocator, reinterpreting as int32_t for the ruy-accumulator.